## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [3]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.dtypes

first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object

In [5]:
df.value_counts

<bound method DataFrame.value_counts of            first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  cli

In [6]:
# What three item categories had the most entries?
top_three_catagories = df['category'].value_counts()
top_three_catagories.head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [7]:
# For the category with the most entries, which subcategory had the most entries?
most_entry_catagory = df['category'].value_counts().idxmax()
filtered_df = df[df['category'] == most_entry_catagory]
top_subcatagory = filtered_df['subcategory'].value_counts(max).idxmax()
top_subcatagory_count = filtered_df['subcategory'].value_counts().max()
print(f"{top_subcatagory}: {top_subcatagory_count}")


bathroom supplies: 6424


In [8]:
# Which five clients had the most entries in the data?
top_5_clients = df['client_id'].value_counts()
top_5_clients.head()

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [9]:
# Store the client ids of those top 5 clients in a list.
stored_top_5_clients =  df['client_id'].value_counts().head(5).index.tolist()
stored_top_5_clients

[33615, 66037, 46820, 38378, 24741]

In [10]:
# How many total units (the qty column) did the client with the most entries order order?
client_with_most_entries = df['client_id'].value_counts().idxmax()
filtered_client_data = df[df['client_id'] == client_with_most_entries]
total_units_ordered = filtered_client_data['qty'].sum()
print(f"Top Client Had A Total of {total_units_ordered} Units Ordered")

Top Client Had A Total of 64313 Units Ordered


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [11]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']
print(df[['unit_price', 'qty', 'subtotal']].head())

   unit_price  qty   subtotal
0     1096.80  105  115164.00
1       24.95   21     523.95
2       13.52   39     527.28
3       36.42   29    1056.18
4      195.10   20    3902.00


In [12]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df['shipping_price'] = (df['unit_weight'] * df['qty']).apply(lambda x: 7 * x if x > 50 else 10 * x)  
print(df[['unit_weight', 'qty', 'shipping_price']].head())


   unit_weight  qty  shipping_price
0         7.50  105         5512.50
1         1.49   21          312.90
2         1.68   39          458.64
3         1.23   29          356.70
4        46.43   20         6500.20


In [13]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
sales_tax = 0.0925
df['total_price'] = (df['subtotal'] + df['shipping_price']) * (1 + sales_tax)
print(df[['shipping_price', 'subtotal', 'total_price']].head())

   shipping_price   subtotal    total_price
0         5512.50  115164.00  131839.076250
1          312.90     523.95     914.258625
2          458.64     527.28    1077.117600
3          356.70    1056.18    1543.571400
4         6500.20    3902.00   11364.403500


In [14]:
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number      subtotal  \
count  54639.000000  54639.000000  5.463900e+04  54639.000000  5.463900e+04   
mean      99.446073      5.004116  5.702646e+02      2.979667  7.710286e+04   
std      133.164267      5.326599  1.879552e+04      2.436320  2.881926e+06   
min        0.010000      0.000000  0.000000e+00      0.000000  0.000000e+00   
25%       14.840000      1.450000  3.200000e+01      1.000000  1.132950e+03   
50%       49.890000      3.240000  6.800000e+01      3.000000  4.629440e+03   
75%      125.570000      6.890000  1.700000e+02      5.000000  1.736913e+04   
max      846.270000     46.430000  3.958244e+06      9.000000  5.844743e+08   

       shipping_price   total_price  
count    5.463900e+04  5.463900e+04  
mean     1.742510e+04  1.032718e+05  
std      3.891595e+05  3.286505e+06  
min      0.000000e+00  0.000000e+00  
25%      4.966500e+02  2.887565e+03  
50%      1.577800e+03  8.396485e+03  
75%      5.029080e+03  2.645605e+04  
max      6.306339e+07  6.412625e+08

In [15]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost'] = (df['unit_cost'] * df['qty']) * df['shipping_price']
print(df[['unit_cost', 'qty', 'shipping_price', 'line_cost']].head())

   unit_cost  qty  shipping_price     line_cost
0     762.71  105         5512.50  4.414661e+08
1      15.09   21          312.90  9.915488e+04
2       7.86   39          458.64  1.405915e+05
3      24.85   29          356.70  2.570559e+05
4     108.17   20         6500.20  1.406253e+07


In [16]:
# Create a column for the profit of each line using line cost and line price
df['line_profit'] = df['total_price'] - df['line_cost']
print(df[['line_cost', 'total_price', 'line_profit']].head())

      line_cost    total_price   line_profit
0  4.414661e+08  131839.076250 -4.413342e+08
1  9.915488e+04     914.258625 -9.824062e+04
2  1.405915e+05    1077.117600 -1.395144e+05
3  2.570559e+05    1543.571400 -2.555123e+05
4  1.406253e+07   11364.403500 -1.405117e+07


In [17]:
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'subtotal', 'shipping_price', 'total_price', 'line_cost',
       'line_profit'],
      dtype='object')

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [18]:
order_ids = [2742071, 2173913, 6128929]
receipt_prices = [152811.89, 162388.71, 923441.25]
calculated_totals = df.groupby('order_id')['total_price'].sum().loc[order_ids].tolist()
order_totals = pd.DataFrame({
    'Order ID': order_ids,
    'Receipt Price': receipt_prices,
    'Calculated Price': calculated_totals
})
order_totals

Order ID  Receipt Price  Calculated Price
0   2742071      152811.89     152811.897075
1   2173913      162388.71     162388.719300
2   6128929      923441.25     923441.240450

In [25]:
import numpy as np

tolerance = 1e-2
order_totals['Match'] = np.isclose(order_totals['Receipt Price'], order_totals['Calculated Price'], atol=tolerance)
for index, row in order_totals.iterrows():
    if row['Match']:
        print(f"Order ID {row['Order ID']} receipt matched the calculated total.")
    else:
        print(f"Order ID {row['Order ID']} receipt did not match the calculated total.")

order_totals

Order ID 2742071 receipt matched the calculated total.
Order ID 2173913 receipt matched the calculated total.
Order ID 6128929 receipt matched the calculated total.


Order ID  Receipt Price  Calculated Price  Match
0   2742071      152811.89     152811.897075   True
1   2173913      162388.71     162388.719300   True
2   6128929      923441.25     923441.240450   True

Ran a tolerance function to eliminate the vibration showing the difference of a penny regarding order id #6128929

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [19]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
if isinstance(stored_top_5_clients, list):
    stored_top_5_clients = pd.DataFrame(stored_top_5_clients)
stored_top_5_clients['total_spent'] = df['unit_price'] * df['qty']
top_5_spending = stored_top_5_clients.groupby(df['client_id'])['total_spent'].sum().reset_index()
top_5_spending

client_id  total_spent
0      37609       523.95
1      46554      1056.18
2      57113       527.28
3      58515    115164.00
4      92089      3902.00

In [20]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
stored_top_5_clients['total_units_purchased'] = df['unit_cost'] * df['qty']
stored_top_5_clients['total_revenue'] = df['unit_price'] * df['qty']
stored_top_5_clients['total_profit'] = (df['unit_price'] - df['unit_cost']) * df['qty']
stored_top_5_clients['total_shipping_price'] = df['unit_weight'] * df['qty']
summary_df = stored_top_5_clients.groupby(df['client_id']).agg(
    total_units_purchased=('total_units_purchased', 'sum'),
    total_shipping_price=('total_shipping_price', 'sum'),
    total_revenue=('total_revenue', 'sum'),
    total_profit=('total_profit', 'sum')
).reset_index()
summary_df


client_id  total_units_purchased  total_shipping_price  total_revenue  \
0      37609                 316.89                 31.29         523.95   
1      46554                 720.65                 35.67        1056.18   
2      57113                 306.54                 65.52         527.28   
3      58515               80084.55                787.50      115164.00   
4      92089                2163.40                928.60        3902.00   

   total_profit  
0        207.06  
1        335.53  
2        220.74  
3      35079.45  
4       1738.60

In [21]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 


# Define a function that converts a dollar amount to millions.


# Apply the currency_format_millions function to only the money columns. 


# Rename the columns to reflect the change in the money format. 
def to_millions(value):
    return value / 1_000_000
summary_df['total_shipping_price'] = summary_df['total_shipping_price'].apply(to_millions)
summary_df['total_revenue'] = summary_df['total_revenue'].apply(to_millions)
summary_df['total_profit'] = summary_df['total_profit'].apply(to_millions)
summary_df = summary_df.rename(columns={
    'client_id': 'Client ID',
    'total_units_purchased': 'Units',
    'total_shipping_price': 'Shipping (millions)',
    'total_revenue': 'Total Revenue (millions)',
    'total_profit': 'Total Profit (millions)'
})
summary_df



Client ID     Units  Shipping (millions)  Total Revenue (millions)  \
0      37609    316.89             0.000031                  0.000524   
1      46554    720.65             0.000036                  0.001056   
2      57113    306.54             0.000066                  0.000527   
3      58515  80084.55             0.000788                  0.115164   
4      92089   2163.40             0.000929                  0.003902   

   Total Profit (millions)  
0                 0.000207  
1                 0.000336  
2                 0.000221  
3                 0.035079  
4                 0.001739

In [22]:
summary_df['Shipping (millions)'] = summary_df['Shipping (millions)'] / 1e6
summary_df['Total Revenue (millions)'] = summary_df['Total Revenue (millions)'] / 1e6
summary_df['Total Profit (millions)'] = summary_df['Total Profit (millions)'] / 1e6
summary_df

Client ID     Units  Shipping (millions)  Total Revenue (millions)  \
0      37609    316.89         3.129000e-11              5.239500e-10   
1      46554    720.65         3.567000e-11              1.056180e-09   
2      57113    306.54         6.552000e-11              5.272800e-10   
3      58515  80084.55         7.875000e-10              1.151640e-07   
4      92089   2163.40         9.286000e-10              3.902000e-09   

   Total Profit (millions)  
0             2.070600e-10  
1             3.355300e-10  
2             2.207400e-10  
3             3.507945e-08  
4             1.738600e-09

In [23]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_summary_df = summary_df.sort_values(by='Total Profit (millions)', ascending=False)
sorted_summary_df

Client ID     Units  Shipping (millions)  Total Revenue (millions)  \
3      58515  80084.55         7.875000e-10              1.151640e-07   
4      92089   2163.40         9.286000e-10              3.902000e-09   
1      46554    720.65         3.567000e-11              1.056180e-09   
2      57113    306.54         6.552000e-11              5.272800e-10   
0      37609    316.89         3.129000e-11              5.239500e-10   

   Total Profit (millions)  
3             3.507945e-08  
4             1.738600e-09  
1             3.355300e-10  
2             2.207400e-10  
0             2.070600e-10